stream/src/main/java/org/apache/kafka/stream/processor/PTopologyBuilder.java

```diff
@@ -0,0 +1,129 @@
+            this.clazz = clazz;
+            this.config = config;
+        }
+    }
+
+    public PTopologyBuilder() {}
+
+    @SuppressWarnings("unchecked")
+    public final KafkaSource addSource(String name, Deserializer keyDeserializer, Deserializer valDeserializer, String... topics) {
+        KafkaSource source = new KafkaSource(name, keyDeserializer, valDeserializer);
+
+        for (String topic : topics) {
+            if (sources.containsKey(topic))
+                throw new IllegalArgumentException("Topic " + topic + " has already been registered by another processor.");
+
+            sources.put(name, source);
+        }
+
+        return source;
+    }
+
+    public final void addProcessor(String name, Class<? extends KafkaProcessor> processorClass, PConfig config, String... parentNames) {
+        if (processorClasses.containsKey(name))
+            throw new IllegalArgumentException("Processor " + name + " is already added.");
+
+        processorClasses.put(name, new ProcessorClazz(processorClass, config));
+
+        if (parentNames != null) {
+            for (String parent : parentNames) {
+                if (!processorClasses.containsKey(parent))
+                    throw new IllegalArgumentException("Parent processor " + parent + " is not added yet.");
+
+                // add to parent list
+                if (!parents.containsKey(name))
+                    parents.put(name, new ArrayList<>());
+                parents.get(name).add(parent);
+
+                // add to parent list
+                if (!children.containsKey(parent))
+                    children.put(parent, new ArrayList<>());
+                children.get(parent).add(name);
+            }
+        }
+    }
+
+    /**
+     * Build the topology by creating the processors
+     */
+    @SuppressWarnings("unchecked")
+    public PTopology build() {
+        Map<String, KafkaProcessor> processorMap = new HashMap<>();
+
+        // create processors
+        try {
+            for (String name : processorClasses.keySet()) {
+                PConfig config = processorClasses.get(name).config;
+                Class<? extends KafkaProcessor> processorClass = processorClasses.get(name).clazz;
+                KafkaProcessor processor = processorClass.getConstructor(String.class, PConfig.class).newInstance(name, config);
+                processorMap.put(name, processor);
+            }
+        } catch (Exception e) {
+            throw new KafkaException("KafkaProcessor(String) constructor failed: this should not happen.");
+        }
+
+        // construct processor parent-child relationships
+        for (String topic : sources.keySet()) {
+            KafkaSource source = sources.get(topic);
+            processorMap.put(source.name(), source);
+
+            // chain children to this processor
+            for (String child : children.get(source.name())) {
+                KafkaProcessor childProcessor = processorMap.get(child);
```

Comment:
@guozhangwang This is less of a childProcessor and more of a sourceProcessor.  Easier to read code if we made that change.